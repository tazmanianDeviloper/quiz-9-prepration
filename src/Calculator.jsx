import { useReducer } from "react";

// Initial state for the calculator
const initialState = {
    firstNumber: 0,
    secondNumber: 0,
    res: 0
};

// Reducer function for performing calculations
function reducer(state, action) {

    switch (action.type) {

        case 'UPDATE_FIRST':
            return {
                ...state,
                firstNumber: parseFloat(action.payload)
            };

        case 'UPDATE_SECOND':
            return {
                ...state,
                secondNumber: parseFloat(action.payload)
            };

        case 'ADDITION':
            return {
                ...state,
                res: state.firstNumber + state.secondNumber
            };

        // Add the other cases here:

        default:
            return state;
    }
}

export default function Calculator() {
    const [state, dispatch] = useReducer(reducer, initialState);

    return (
        <>
            <label>Enter a number: </label>
            <input
                id="first"
                type="number"
                value={state.firstNumber}
                onChange={(e) => dispatch({ type: 'UPDATE_FIRST', payload: e.target.value })}
            />

            <label>Enter another number: </label>
            <input
                id="second"
                type="number"
                value={state.secondNumber}
                onChange={(e) => dispatch({ type: 'UPDATE_SECOND', payload: e.target.value })}
            />

            {/*Addition Button*/}
            <button onClick={() => dispatch({type: 'ADDITION'})}>
                +
            </button>

            {/*Add Subtraction Button here*/}


            {/*Add Multiplication Button here*/}


            {/*Add Division Button here*/}


            <p>Result: {state.res}</p>
        </>
    );
}

