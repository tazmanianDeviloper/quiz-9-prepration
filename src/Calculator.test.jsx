import Calculator from "./Calculator.jsx";
import renderer from 'react-test-renderer';
import {test, expect} from 'vitest';

test('The Calculator component should run', ()=>{
    const view = renderer
        .create(<Calculator/>)
        .toJSON();
    expect(view).toMatchSnapshot()
});

test('Adding two Numbers should result in their summation', ()=>{
    const component = renderer.create(<Calculator/>);
    const firstInput = component.root.findByProps({ id: 'first' });
    const secondInput = component.root.findByProps({ id: 'second' });
    const addButton = component.root.findAllByType('button')[0];

    // Change input values
    firstInput.props.onChange({ target: { value: '2' } });
    secondInput.props.onChange({ target: { value: '3' } });

    // Click addition button
    addButton.props.onClick();

    // Find result
    const resultArray = component.root.findByType('p').props.children;

    // Extract the numerical result
    const resultValue = resultArray[1];

    // Assert result is correct
    expect(resultValue).toBe(5);
});

test('Subtracting two numbers should display the correct difference', () => {
    const component = renderer.create(<Calculator />);
    const firstInput = component.root.findByProps({ id: 'first' });
    const secondInput = component.root.findByProps({ id: 'second' });
    const subtractButton = component.root.findAllByType('button')[1];

    // Change input values
    firstInput.props.onChange({ target: { value: '8' } });
    secondInput.props.onChange({ target: { value: '3' } });

    // Click subtraction button
    subtractButton.props.onClick();

    // Find result
    const resultArray = component.root.findByType('p').props.children;

    // Extract the numerical result
    const resultValue = resultArray[1];

    // Assert result is correct
    expect(resultValue).toBe(5);
});

test('Multiplying two numbers should display the correct product', () => {
    const component = renderer.create(<Calculator />);
    const firstInput = component.root.findByProps({ id: 'first' });
    const secondInput = component.root.findByProps({ id: 'second' });
    const multiplyButton = component.root.findAllByType('button')[2];

    // Change input values
    firstInput.props.onChange({ target: { value: '4' } });
    secondInput.props.onChange({ target: { value: '3' } });

    // Click multiplication button
    multiplyButton.props.onClick();

    // Find result
    const resultArray = component.root.findByType('p').props.children;

    // Extract the numerical result
    const resultValue = resultArray[1];

    // Assert result is correct
    expect(resultValue).toBe(12);
});

test('Dividing two numbers should display the correct quotient', () => {
    const component = renderer.create(<Calculator />);
    const firstInput = component.root.findByProps({ id: 'first' });
    const secondInput = component.root.findByProps({ id: 'second' });
    const divideButton = component.root.findAllByType('button')[3];

    // Change input values
    firstInput.props.onChange({ target: { value: '10' } });
    secondInput.props.onChange({ target: { value: '2' } });

    // Click division button
    divideButton.props.onClick();

    // Find result
    const resultArray = component.root.findByType('p').props.children;

    // Extract the numerical result
    const resultValue = resultArray[1];

    // Assert result is correct
    expect(resultValue).toBe(5);
});
