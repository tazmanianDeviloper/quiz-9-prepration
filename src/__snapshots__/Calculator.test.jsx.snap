// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`The Calculator component should run 1`] = `
[
  <label>
    Enter a number: 
  </label>,
  <input
    id="first"
    onChange={[Function]}
    type="number"
    value={0}
  />,
  <label>
    Enter another number: 
  </label>,
  <input
    id="second"
    onChange={[Function]}
    type="number"
    value={0}
  />,
  <button
    onClick={[Function]}
  >
    +
  </button>,
  <button
    onClick={[Function]}
  >
    -
  </button>,
  <button
    onClick={[Function]}
  >
    *
  </button>,
  <button
    onClick={[Function]}
  >
    /
  </button>,
  <p>
    Result: 
  </p>,
]
`;
